### PR TITLE
fix(approval): contextual exemption for safe compound command patterns

### LINKED ISSUE #58: bug(approval): contextual exemption cannot handle safe compound command patterns
## Bug

Legitimate compound command patterns commonly used with `gh api` are sent to Haiku for approval even when the `gh api` portion targets a related repo with an allowed method. Unlike #57 (false positives on quoted arguments), these commands genuinely contain shell operators — but the overall intent is safe.

### Reproduction

```
gh api repos/perezd/claudetainer/issues/comments/4174499209 -X PATCH --input <(jq -n --rawfile body /tmp/plan-updated.md '{"body": $body}') 2>&1 | head -5
```

This triggers approval because:
- `<(...)` — process substitution
- `| head -5` — shell pipe
- `2>&1` — redirection (contains `&`)

All are real shell operators correctly detected by `hasCompoundOperators`, but the `gh api` portion is a simple PATCH to a related repo.

### Common safe compound patterns

- `gh api ... | head -N` / `| tail -N` / `| jq ...` — output filtering
- `gh api ... 2>&1` — stderr redirection
- `gh api --input <(jq ...)` — process substitution for request body
- `cd /workspace/repo && gh api ...` — directory change before API call

### Why this is harder than #57

#57 is about metacharacters inside quoted arguments (false positives). This issue involves real shell operators where the approval system would need to understand the *structure* of the compound command — extract the `gh api` segment, validate it independently, and verify the other segments are benign. This has a larger security surface and requires more careful design.

### Relationship to #57

These are independent fixes. #57 (quoted argument false positives) should be addressed first as it's simpler and higher-impact. This issue can be tackled separately.

### PR DESCRIPTION
**Pull request by limbibot**
## Summary

- Add `extractCoreCommand()` function that strips known-safe shell wrappers (stderr redirection `2>&1`, trailing pipes to read-only filters, leading `cd` prefix) from commands before the `hasCompoundOperators` check in the contextual exemption path
- Close pre-existing gap: add `<` and `>` (file redirection operators) to `COMPOUND_OPERATORS_RE` to prevent bypass via output redirection
- Integrate into `isContextualGhCommand()` with distinct `core-extracted` logging tag for forensic differentiation
- Expand CLAUDE.md CLI best practices to require temp files for all `gh` body/payload arguments (`--body-file` for high-level commands, `--input` for `gh api`)

Closes #58

## Layer-Impact Assessment

1. **Container Hardening** — Not affected. No changes to user namespaces, rootfs, tmpfs, or capabilities.
2. **Network Isolation** — Not affected. No changes to iptables, DNS, or domain allowlist.
3. **Command Approval** — Affected. Refines the contextual exemption path in `isContextualGhCommand()` to handle safe compound patterns without sending them to Haiku. Tier 1 hard-blocks are unchanged (still see full command). Haiku remains the fallback for unrecognized patterns.

Full panel review triggered and completed (2 rounds).

## Security Design Checklist

- **Trust anchor mutability:** No. Safe-filter allowlist is a compile-time `Set<string>` constant. No runtime-writable data sources.
- **File and output visibility:** N/A — no files created. Existing `[HOOK]` stderr logging extended with distinct `core-extracted` tag.
- **Allowlist vs blocklist:** Allowlist. Safe wrappers and safe pipe targets are explicitly enumerated. Unknown patterns fail-closed to Haiku.
- **Fail mode:** Fail-closed. Unrecognized wrappers remain in the string → `hasCompoundOperators` catches them → Haiku. Pipe filter args that look like file paths (starting with `/`, `~`, `.`, or quoted equivalents) are rejected.
- **Temporal safety:** N/A — pure string processing in the same process/context. No privilege transitions, no TOCTOU window.
- **Network exposure:** N/A — no new network communication. Reduces unnecessary Haiku API calls.
- **Layer compensation:** No layer weakened. Tier 1 still sees the full unmodified command. Network isolation operates independently.

## Panel Review

| Expert | Round 1 | Round 2 |
|--------|---------|---------|
| Container Security | **Approve** | — |
| Cloud Infrastructure | Approve-with-conditions | **Approve** |
| Red Team | Request-changes | **Approve-with-conditions** |
| Compliance & Risk | Approve-with-conditions | **Approve** |

Round 1 findings addressed:
- Added `<>` to `COMPOUND_OPERATORS_RE` (red team: high severity)
- Removed process substitution stripping (red team: high severity — hides dangerous subcommands)
- Added distinct `core-extracted` logging tag (compliance)

Round 2: All experts approved. Red team condition (verify trailing pipe rejects chained operators, cd path metachar check covers `$`/backticks/parens) verified in implementation.

Final code review finding addressed: reject pipe filter args that look like file paths.

## PR Review Findings Addressed

Copilot review (4 findings, all addressed in `17ba83f`):
1. **Quoted file paths bypass `hasPathArg`** — Updated regex to catch `"`, `'` before path prefixes
2. **Greedy pipe regex splits on `|` inside single quotes** — Replaced with `findLastUnquotedPipe()` character scanner that tracks quote state
3. **Relative filenames bypass path check** — Removed `cat` from `SAFE_PIPE_FILTERS` (remaining filters only consume stdin)
4. **Missing edge case tests** — Added 5 tests: quoted paths (3), jq with internal `|` (2)

## Test Plan

- [x] 392 tests pass, 0 failures
- [x] `extractCoreCommand` — 32 unit tests covering stderr redirection, trailing safe pipes, chained pipes, unsafe pipe targets, file-path arg rejection (including quoted paths), cd prefix with clean/dirty paths, process substitution preservation, jq with internal `|`, combinations
- [x] `hasCompoundOperators` — 5 new tests for `<`, `>`, `>>` detection and single-quote exemption
- [x] `isContextualGhCommand` — 9 new integration tests for compound wrapper exemption and rejection cases
- [x] Issue #58 reproduction cases — 3 end-to-end tests using exact commands from the issue
- [x] Prettier formatting verified

### CODE REVIEW

File: approval/check-command.ts [hunk 1]
```diff
@@ -348,6 +348,80 @@ export function hasCompoundOperators(command: string): boolean {
...
+      // a) The filter command is in the safe allowlist
+      // b) The filter args contain no further shell operators
+      // c) No args look like file paths (safe filters can read files
+      //    when given filename args, bypassing the stdin-only assumption)
+      const strippedArgs = filterArgs.replace(SINGLE_QUOTE_PAIR_RE, "");
+      const hasPathArg = /(?:^|\s)[/~.]/.test(strippedArgs);
+      if (
+        SAFE_PIPE_FILTERS.has(filterCmd) &&
+        !COMPOUND_OPERATORS_RE.test(strippedArgs) &&
+        !hasPathArg
```
[the remark this answers is not in the document]
Comment (limbibot) [reply approval/check-command.ts]: Fixed. Updated `hasPathArg` regex to `/(?\:^|\s)["']?[/~.]/` so it catches `"/etc/passwd"`, `"./config"`, etc. Also removed `cat` from `SAFE_PIPE_FILTERS` entirely (see comment 3 reply).

File: approval/check-command.ts [hunk 2]
```diff
@@ -348,6 +348,80 @@ export function hasCompoundOperators(command: string): boolean {
...
+  cmd = cmd.replace(/\s*2>&1\s*/g, " ").trim();
+
+  // 3. Strip trailing pipe to safe filter (repeated for chains)
+  let changed = true;
+  while (changed) {
+    changed = false;
+    // Match the last `| <cmd> [args]` segment (greedy .* to find the last pipe)
+    const pipeMatch = cmd.match(/^(.*)\s*\|\s*(\S+)(.*)$/);
+    if (pipeMatch) {
+      const [, before, filterCmd, filterArgs] = pipeMatch;
```
[the remark this answers is not in the document]
Comment (limbibot) [reply approval/check-command.ts]: Fixed. Replaced the greedy regex with `findLastUnquotedPipe()` — a character-by-character scan that tracks single-quote state to find the last unquoted `|`. Now `jq '.items | .id'` correctly resolves to `jq` as the filter command, not `.id'`.

File: approval/check-command.ts [hunk 3]
```diff
@@ -348,6 +348,80 @@ export function hasCompoundOperators(command: string): boolean {
...
+      // b) The filter args contain no further shell operators
+      // c) No args look like file paths (safe filters can read files
+      //    when given filename args, bypassing the stdin-only assumption)
+      const strippedArgs = filterArgs.replace(SINGLE_QUOTE_PAIR_RE, "");
+      const hasPathArg = /(?:^|\s)[/~.]/.test(strippedArgs);
+      if (
+        SAFE_PIPE_FILTERS.has(filterCmd) &&
+        !COMPOUND_OPERATORS_RE.test(strippedArgs) &&
+        !hasPathArg
+      ) {
```
[the remark this answers is not in the document]
Comment (limbibot) [reply approval/check-command.ts]: Removed `cat` from `SAFE_PIPE_FILTERS`. The `| cat` no-args passthrough case is not worth the risk of `| cat secrets.txt` with relative paths that evade path detection. Remaining filters (`head`, `tail`, `jq`, `wc`, `grep`) only read stdin in normal pipeline usage.

File: approval/check-command.ts [hunk 4]
```diff
@@ -348,6 +348,105 @@ export function hasCompoundOperators(command: string): boolean {
...
+      // b) The filter args contain no further shell operators
+      // c) No args look like file paths (safe filters can read files
+      //    when given filename args, bypassing the stdin-only assumption)
+      const strippedArgs = filterArgs.replace(SINGLE_QUOTE_PAIR_RE, "");
+      const hasPathArg = /(?:^|\s)["']?[/~.]/.test(strippedArgs);
+      if (
+        SAFE_PIPE_FILTERS.has(filterCmd) &&
+        !COMPOUND_OPERATORS_RE.test(strippedArgs) &&
+        !hasPathArg
+      ) {
```
[the remark this answers is not in the document]
Comment (limbibot) [reply approval/check-command.ts]: Fixed. hasPathArg now runs on original filterArgs instead of strippedArgs. Also refined the regex to require . be followed by / so jq expressions like .field are not falsely flagged as paths.
